pkg/p2p/libp2p/libp2p.go

```diff
@@ -884,7 +884,19 @@ func (s *Service) Ping(ctx context.Context, addr ma.Multiaddr) (rtt time.Duratio
 // provides it. It ignores the default libp2p user agent string
 // "github.com/libp2p/go-libp2p" and returns empty string in that case.
 func (s *Service) peerUserAgent(peerID libp2ppeer.ID) string {
-	v, err := s.host.Peerstore().Get(peerID, "AgentVersion")
+	var (
+		v   interface{}
+		err error
+	)
+	// Peerstore may not contain all keys and values right after the connections is created.
+	// This retry mechanism ensures more reliable user agent propagation.
+	for deadline := time.Now().Add(2 * time.Second); time.Now().Before(deadline); {
```

Comment:
my only question would be: could it be that the retry mechanism would make some context with deadline in the enclosing context expire?
example:
```go
func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr) error {
	ctx, cancel := context.WithTimeout(ctx, peerConnectionAttemptTimeout)
	defer cancel()

	switch i, err := k.p2p.Connect(ctx, ma); {
```
kademlia dial has a 5 seconds timeout in this case. so the user agent printing might delay this and make it expire. I wonder if we could either decrease this timespan to be shorter so that it becomes insignificant comparing to the total connection context deadline or simply do this polling in a different goroutine and print it eventually in an unrelated logline that would have the overlay and user agent together.

Reply:
Good suggestion. I've added context handling to the function.